tree/ntuple/v7/inc/ROOT/RNTuple.hxx

```diff
@@ -378,6 +378,68 @@ private:
    /// Estimator of uncompressed cluster size, taking into account the estimated compression ratio
    std::size_t fUnzippedClusterSizeEst;
 
+public:
+   RNTupleFillContext(std::unique_ptr<RNTupleModel> model, std::unique_ptr<Detail::RPageSink> sink);
+   RNTupleFillContext(const RNTupleFillContext &) = delete;
+   RNTupleFillContext &operator=(const RNTupleFillContext &) = delete;
+   ~RNTupleFillContext();
+
+   /// Fill an entry into this context.  This method will perform a light check whether the entry comes from the
+   /// context's own model.
+   /// \return The number of uncompressed bytes written.
+   std::size_t Fill(REntry &entry)
+   {
+      if (R__unlikely(entry.GetModelId() != fModel->GetModelId()))
+         throw RException(R__FAIL("mismatch between entry and model"));
+
+      const std::size_t bytesWritten = entry.Append();
+      fUnzippedClusterSize += bytesWritten;
+      fNEntries++;
+      if ((fUnzippedClusterSize >= fMaxUnzippedClusterSize) || (fUnzippedClusterSize >= fUnzippedClusterSizeEst))
+         CommitCluster();
+      return bytesWritten;
+   }
+   /// Ensure that the data from the so far seen Fill calls has been written to storage
+   void CommitCluster();
+
+   std::unique_ptr<REntry> CreateEntry() { return fModel->CreateEntry(); }
+
+   /// Return the entry number that was last committed in a cluster.
+   NTupleSize_t GetLastCommitted() const { return fLastCommitted; }
+   /// Return the number of entries filled so far.
+   NTupleSize_t GetNEntries() const { return fNEntries; }
+
+   void EnableMetrics() { fMetrics.Enable(); }
+   const Detail::RNTupleMetrics &GetMetrics() const { return fMetrics; }
+};
+
+// clang-format off
+/**
+\class ROOT::Experimental::RNTupleWriter
+\ingroup NTuple
+\brief An RNTuple that gets filled with entries (data) and writes them to storage
+
+An output ntuple can be filled with entries. The caller has to make sure that the data that gets filled into an ntuple
+is not modified for the time of the Fill() call. The fill call serializes the C++ object into the column format and
+writes data into the corresponding column page buffers.  Writing of the buffers to storage is deferred and can be
+triggered by CommitCluster() or by destructing the writer.  On I/O errors, an exception is thrown.
+*/
+// clang-format on
+class RNTupleWriter {
+   friend RNTupleModel::RUpdater;
+
+private:
+   /// The page sink's parallel page compression scheduler if IMT is on.
+   /// Needs to be destructed after the page sink (in the fill context) is destructed and so declared before.
+   std::unique_ptr<Detail::RPageStorage::RTaskScheduler> fZipTasks;
+   RNTupleFillContext fFillContext;
+   Detail::RNTupleMetrics fMetrics;
+
+   NTupleSize_t fLastCommittedClusterGroup = 0;
+
+   RNTupleModel &GetUpdatableModel() { return *fFillContext.fModel; }
+   Detail::RPageSink &GetSink() { return *fFillContext.fSink; }
```

Comment:
These two are needed for `RNTupleModel::RUpdater` - the alternative would be to declare `friend RNTupleModel::RUpdater` also in the `RNTupleFillContext` which I don't like very much...